src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -25,10 +25,29 @@
 from gt4py.next.utils import flatten_nested_tuple, tree_map
 
 
-DOMAIN: TypeAlias = domain_utils.SymbolicDomain | None | tuple["DOMAIN", ...]
+class DomainAccessDescriptor(eve.StrEnum):
+    """
+    Descriptor for domains that could not be inferred.
+    """
+
+    #: The access if unknown because of a dynamic shift.whose extent is not known.
+    #: E.g.: `(⇑(λ(arg0, arg1) → ·⟪Ioffₒ, ·arg1⟫(arg0)))(in_field1, in_field2)`
+    UNKNOWN = "unknown"
+    #: The domain is never accessed.
+    #: E.g.: `{in_field1, in_field2}[0]`
+    NEVER = "never"
+
+
+DOMAIN: TypeAlias = domain_utils.SymbolicDomain | DomainAccessDescriptor | tuple["DOMAIN", ...]
```

Comment:
```suggestion
_Domain: TypeAlias = domain_utils.SymbolicDomain | DomainAccessDescriptor | tuple["Domain", ...]
```

Reply:
They can be used from the outside and this is also desired if you use `infer_expr` directly, so I left them public.

Reply:
Where do you do that?

Reply:
Here: https://github.com/GridTools/gt4py/pull/1738/files#diff-36c97ae1e7197fbdd81532e533a2e6294febc1c23bbb1a6ef2968130e15368fbR77
I added an annotation there so that we can reverse search.